src/dials/algorithms/indexing/indexer.py

```diff
@@ -613,6 +613,11 @@ def index(self):
 
                 if self._check_have_similar_crystal_models(experiments):
                     have_similar_crystal_models = True
+                    try:
+                        # Unset ids for the last experiment that is now deleted
+                        self._remove_id_from_reflections(len(experiments))
```

Comment:
I think this may be overlooking the case of shared crystal models.  A possibility is that you have four experiments and two crystals (e.g. multi-sweep indexing). I think case you need to match experiments that share the crystal to be removed, which is what is done in the other case of `_remove_id_from_reflections` below.

Reply:
I see. I think the additional shared crystal matching could be incorporated into the function so there's one place that does the removal properly. However, this is an area where I've run into trouble before (https://github.com/dials/dials/issues/2609). Maybe there are use-cases where we don't want to throw out all experiments if the crystal model is bad for just one of them. This might need more thought...